File: framework/environment/src/main/java/top/bettercode/environment/ConfigEndpoint.java
package top.bettercode.environment;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import org.springframework.boot.actuate.endpoint.annotation.Endpoint;
import org.springframework.boot.actuate.endpoint.annotation.ReadOperation;
import org.springframework.boot.env.YamlPropertySourceLoader;
import org.springframework.core.env.MapPropertySource;
import org.springframework.core.env.PropertySource;
import org.springframework.core.io.UrlResource;
import org.springframework.lang.NonNull;
import top.bettercode.lang.util.StringUtil;

/**
 * @author Peter Wu
 */
@Endpoint(
    id = "config"
)
public class ConfigEndpoint {

  private final EnvironmentManager environment;

  public ConfigEndpoint(EnvironmentManager environment) {
    this.environment = environment;
  }

  @ReadOperation
  public Object write(@NonNull String ymlUrl) {
    try {
      Map<String, String> changed = new HashMap<>();
      List<PropertySource<?>> propertySources = new YamlPropertySourceLoader().load("remote",
          new UrlResource(ymlUrl));
      for (PropertySource<?> propertySource : propertySources) {
        MapPropertySource mapPropertySource = (MapPropertySource) propertySource;
        for (Entry<String, Object> entry : mapPropertySource.getSource().entrySet()) {
          Object value = entry.getValue();
          String key = entry.getKey();
          String val = value == null ? null : value.toString();
          boolean change = this.environment.setProperty(key, val);
          if (change) {
            changed.put(key, val);
          }
        }
      }
      return changed;
    } catch (IOException e) {
      return "配置文件加载错误：" + "\n" + StringUtil.valueOf(e);
    }
  }
}

File: env/src/main/kotlin/top/bettercode/summer/env/ConfigEndpoint.kt
package top.bettercode.summer.env

import org.slf4j.LoggerFactory
import org.springframework.boot.actuate.endpoint.annotation.Endpoint
import org.springframework.boot.actuate.endpoint.annotation.ReadOperation
import org.springframework.core.env.MapPropertySource
import top.bettercode.summer.env.RemoteEnvironmentPostProcessor.Companion.loadConfig
import java.io.IOException

/**
 * @author Peter Wu
 */
@Endpoint(id = "config")
class ConfigEndpoint(private val manager: EnvironmentManager) {

    private val log = LoggerFactory.getLogger(ConfigEndpoint::class.java)

    @ReadOperation
    fun write(): Any {
        return try {
            val changed: MutableMap<String, String?> = HashMap()
            loadConfig(manager.environment)?.let {
                it.second.forEach { propertySource ->
                    val mapPropertySource = propertySource as MapPropertySource
                    for ((key, value) in mapPropertySource.source) {
                        val `val` = value?.toString()
                        val change = manager.setProperty(key, `val`)
                        if (change) {
                            changed[key] = `val`
                        }
                    }
                }
                if (log.isInfoEnabled) {
                    log.info("load config in:${it.first.uri}/${it.first.configFile}")
                }
            }
            changed
        } catch (e: IOException) {
            "配置文件加载错误：\n${e.stackTraceToString()}"
        }
    }
}
